### ISSUE TITLE
unstable units processing

### ISSUE THREAD

**Issue by sumkincpp** [NLnetLabs/rtrtr#86]
We have simple setup of units `local -> slurm -> rtrtr`.

Unfortunately when starting rtrtr

- rtrtr may hang and not listen for queries
- slurm may not apply it's exceptions

# Setup

`RTRTR 0.2.2`

## roas.json

```
{
    "roas": [
        {
            "prefix": "192.0.2.1/32",
            "maxLength": 32,
            "asn": "AS0",
            "ta": "TEST-NET-1"
        }      
    ]
}
```

## slurm.json

```
{
  "slurmVersion": 1,
  "validationOutputFilters": {
    "prefixFilters": [],
    "bgpsecFilters": []
  },
  "locallyAddedAssertions": {
    "prefixAssertions": [
      {
        "asn": 0,
        "prefix": "192.0.2.2/32",
        "maxPrefixLength": 32,
        "comment": "RTR Health Check"
      }
    ],
    "bgpsecAssertions": []
  }
}
```

## rtr.conf

```
log_level = "debug"
log_target = "stderr"
http-listen = ["0.0.0.0:8090"]

[units.json]
type = "json"
uri = "file:/test/roas.json"
refresh = 60

[units.slurm]
type = "slurm"
source = "json"
files = [ "/test/slurm.json" ]

[targets.local-9001]
type = "rtr"
listen = [ "0.0.0.0:9001" ]
unit = "slurm"

[targets.http-8090]
type = "http"
path = "/json"
format = "json"
unit = "slurm"
```

# Case 0: normal operation

```
HTTP server listening on 0.0.0.0:8090
Target http-8090: link status: healthy
Unit json: successfully updated.
Target http-8090: Got update (2 entries)
Target local-9001: Got update (2 entries)
Target http-8090: link status: healthy
```

```bash
$ rtrclient -e -t csv -o /dev/stdout tcp 127.0.0.1 9001 2>/dev/null
Sync done
192.0.2.1, 32, 32, 0
192.0.2.2, 32, 32, 0
```

# Case 1: rtrtr hangs

When it hangs on start, no data can be fetched through rtrtr protocol

```
HTTP server listening on 0.0.0.0:8090
Target http-8090: link status: healthy
Unit json: successfully updated.
RTR: Got reset query.
```

```bash
$ rtrclient -e -t csv -o /dev/stdout tcp 127.0.0.1 9001 2>/dev/null
^C
```

# Case 2: 

Slurm exceptions are not applied at start:

```
HTTP server listening on 0.0.0.0:8090
Target http-8090: link status: healthy
Unit json: successfully updated.
Target local-9001: Got update (1 entries)
Target http-8090: Got update (1 entries)
Target http-8090: link status: healthy
RTR: Got reset query.
Unit json: update without changes.
Unit json: update without changes.
RTR: Got reset query.
Unit json: update without changes.
```

As long as nothing changes, slurm contents won't appear

```bash
$ rtrclient -e -t csv -o /dev/stdout tcp 127.0.0.1 9001 2>/dev/null
Sync done
192.0.2.1, 32, 32, 0

$ rtrclient -e -t csv -o /dev/stdout tcp 127.0.0.1 9001 2>/dev/null
Sync done
192.0.2.1, 32, 32, 0
```

**Comment by partim** [comment]
Thank you for the report!

I suspect the Slurm files not applied upon startup is a race condition between loading the payload JSON and the Slurm JSON – I will fix it.

For the case of RTRTR hanging upon start, I can’t seem to reproduce this here. How often does it happen? What environment are you running it in, i.e., real machine or virtual machine?

**Comment by sumkincpp** [comment]
I've built a docker image and use it in virtual machine for testing:
rtrtr hangs maybe in 5-10% cases - i was able to reproduce it in 3-4 cases, but this was hard to catch this moment.

**Comment by partim** [comment]
Ok, I’ll have to keep trying to reproduce the issue, then.

**Comment by sumkincpp** [comment]
Hi @partim,
Thank you very much for the quick fix, can't wait to test it with a new release!

**Comment by partim** [comment]
I still haven’t reproduced the hang-on-start issue. Could you perchance try and see if this also happens if you leave out the Slurm unit?

**Comment by partim** [comment]
In case you still have this issue: I have since figured out the startup issues and hope to have them fixed through PR #101.
